File: repos/farm-server/src/main/com/topcoder/services/util/LongTesterIO.java
/*
* Copyright (C) - 2014 TopCoder Inc., All Rights Reserved.
*/
package com.topcoder.services.util;
import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.PrintStream;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

/**
 *
 * <p>
 * Changes in version 1.1 (Return Peak Memory Usage for Marathon Match Java v1.0):
 * <ol>
 *      <li>Add {@link #PEAK_MEMORY_USED} field.</li>
 *      <li>Add {@link #writePeakMemoryUsed(BufferedOutputStream bw, long peakMemoryUsed)} method.</li>
 * </ol>
 * </p>
 * @author TCSASSEMBLER
 * @version 1.1
 *
 */
public class LongTesterIO{
    static Object resultObject;
    public static final int INT_TYPE = 1;
    public static final int LONG_TYPE = 2;
    public static final int DOUBLE_TYPE = 3;
    public static final int STRING_TYPE = 4;
    public static final int VEC_INT_TYPE = 5;
    public static final int VEC_LONG_TYPE = 6;
    public static final int VEC_DOUBLE_TYPE = 7;
    public static final int VEC_STRING_TYPE = 8;

    public static final int EXCEPTION = 250;
    public static final int TIMEOUT = 251;
    public static final int ABORT = 252;
    public static final int METHOD_START = 253;
    public static final int TIME = 254;
    public static final int TERMINATE = 255;

    /**
     * The Peak Memory Used Type.
     * @since 1.1
     */
    public static final int PEAK_MEMORY_USED = 245;
    
    public static void writeInt(BufferedOutputStream bw, int a) throws IOException{
        bw.write(a>>>24);
        bw.write((a>>>16)&0xff);
        bw.write((a>>>8)&0xff);
        bw.write(a&0xff);
    }
    public static void writeLong(BufferedOutputStream bw, long a) throws IOException{
        bw.write((int)(a>>>56));
        bw.write((int)((a>>>48)&0xff));
        bw.write((int)((a>>>40)&0xff));
        bw.write((int)((a>>>32)&0xff));
        bw.write((int)((a>>>24)&0xff));
        bw.write((int)((a>>>16)&0xff));
        bw.write((int)((a>>>8)&0xff));
        bw.write((int)(a&0xff));
    }
    public static void writeDouble(BufferedOutputStream bw, double a) throws IOException{
        writeLong(bw,Double.doubleToRawLongBits(a));//XXX is Raw faster?
    }
    public static void writeString(BufferedOutputStream bw, String a) throws IOException{
        writeInt(bw,a.length());
        bw.write(a.getBytes());
    }

    public static void setTimeout(BufferedOutputStream bw, int timeout) throws IOException {
        bw.write(TIMEOUT);
        writeInt(bw, timeout);
    }
    
    public static void startMethod(BufferedOutputStream bw, int method) throws IOException {
        bw.write(METHOD_START);
        writeInt(bw, method);
    }
    
    //sends terminate command to close client while loop
    public static void terminate(BufferedOutputStream bw) throws IOException {
        bw.write(TERMINATE);
        flush(bw);
        //socket.close();
    }
    
    public static void shutdownSocket() throws IOException {
        safeClose(client);
        safeClose(readClient);
        safeClose(socket);
        safeClose(readSocket);
    }

    private static void safeClose(Socket s) throws IOException {
        if (s != null) {
            s.close();
        }
    }
    
    private static void safeClose(ServerSocket s) throws IOException {
        if (s != null) {
            s.close();
        }
    }
    public static void flush(BufferedOutputStream bw) throws IOException {
        bw.flush();
    }
    
    public static void writeTime(BufferedOutputStream bw, int time) throws IOException {
        bw.write(TIME);
        writeInt(bw,time);
    }
    /**
     * Write peak memory used.
     * @param bw the output stream.
     * @param peakMemoryUsed the peak memory used value.
     * @throws IOException if any io error occur.
     * @since 1.1
     */
    public static void writePeakMemoryUsed(BufferedOutputStream bw, long peakMemoryUsed) throws IOException {
        bw.write(PEAK_MEMORY_USED);
        writeLong(bw,peakMemoryUsed);
    }
    
    public static void writeThrowable(BufferedOutputStream bw, Throwable t) throws IOException {
        bw.write(EXCEPTION);
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        PrintStream ps = new PrintStream(baos);
        t.printStackTrace(ps);
        String s = new String(baos.toByteArray());
        System.err.println("Writing exception " + s);
        //System.err.println("java.home=" + System.getProperty("java.home"));
        writeString(bw,s);
    }
    
    //for things like timeouts in plain text
    public static void writeThrowable(BufferedOutputStream bw, String s) throws IOException {
        bw.write(EXCEPTION);
        writeString(bw,s);
    }
    
    public static void writeArg(BufferedOutputStream bw, double a) throws IOException{
        bw.write(DOUBLE_TYPE);
        writeDouble(bw,a);
    }
    public static void writeArg(BufferedOutputStream bw, int a) throws IOException{
        bw.write(INT_TYPE);
        writeInt(bw,a);
    }
    public static void writeArg(BufferedOutputStream bw, long a) throws IOException{
        bw.write(LONG_TYPE);
        writeLong(bw,a);
    }
    public static void writeArg(BufferedOutputStream bw, String a) throws IOException{
        bw.write(STRING_TYPE);
        writeString(bw,a);
    }
    public static void writeArg(BufferedOutputStream bw, int[] a) throws IOException{
        bw.write(VEC_INT_TYPE);
        writeInt(bw,a.length);
        for(int i = 0; i<a.length; i++){
            writeInt(bw,a[i]);
        }
    }
    public static void writeArg(BufferedOutputStream bw, long[] a) throws IOException{
        bw.write(VEC_LONG_TYPE);
        writeInt(bw,a.length);
        for(int i = 0; i<a.length; i++){
            writeLong(bw,a[i]);
        }
    }
    public static void writeArg(BufferedOutputStream bw, double[] a) throws IOException{
        bw.write(VEC_DOUBLE_TYPE);
        writeInt(bw,a.length);
        for(int i = 0; i<a.length; i++){
            writeDouble(bw,a[i]);
        }
    }
    public static void writeArg(BufferedOutputStream bw, String[] a) throws IOException{
        bw.write(VEC_STRING_TYPE);
        writeInt(bw,a.length);
        for(int i = 0; i<a.length; i++){
            writeString(bw,a[i]);
        }
    }
    
    public static void initialize(int port) throws IOException {
        exceptions = new ArrayList();
        resultObject = null;
        time = 0;
        
        //bind the socket
        socket = new ServerSocket(port,10, InetAddress.getByName(null));
        socket.setSoTimeout(30000);
        
        readSocket = new ServerSocket(port+1,10, InetAddress.getByName(null));
        readSocket.setSoTimeout(30000);
    }
    
    static Socket client;
    static Socket readClient;
    
    public static Socket getSocket() throws IOException {
        client = socket.accept();
        client.setTcpNoDelay(true);
        return client;
    }
    
    public static Socket getReadSocket() throws IOException {
        readClient = readSocket.accept();
        readClient.setTcpNoDelay(true);
        return readClient;
    }
    
    private static ServerSocket socket = null;
    
    private static ServerSocket readSocket = null;
    
    private static List exceptions = new ArrayList();
    private static long time = 0;
    
    public static List getExceptions() {
        return exceptions;
    }
    
    public static void setTime(long t) {
        time = t;
    }
    
    public static void addTime(long t) {
        time += t;
    }
    
    public static void addFatalError(String s) {
        exceptions.add(s);
    }
    
    public static void setResultObject(Object o) {
        resultObject = o;
    }
    
    public static Object getResultObject() {
        return resultObject;
    }
    
    public static long getTime() {
        return time;
    }
    public static int getInt2(BufferedInputStream br) throws IOException{
        int r = br.read() << 24;
        r |= br.read()<<16;
        r |= br.read()<<8;
        r |= br.read();
        return r;
    }
    //XXX this is sucky, is string buffer faster, less mem?
    public static String getString2(BufferedInputStream br) throws IOException{
        int length = getInt2(br);
        byte[] b = new byte[length];
        for(int i = 0; i<length; i++){
            b[i] = (byte)br.read();
        }
        return new String(b);
    }
    public static long getLong2(BufferedInputStream br) throws IOException{
        long r = ((long)br.read()) << 56;
        r |= ((long)br.read())<<48;
        r |= ((long)br.read())<<40;
        r |= ((long)br.read())<<32;
        r |= ((long)br.read())<<24;
        r |= ((long)br.read())<<16;
        r |= ((long)br.read())<<8;
        r |= ((long)br.read());
        return r;
    }
    public static double getDouble2(BufferedInputStream br) throws IOException{
        return Double.longBitsToDouble(getLong2(br));
    }
    
    public static int getInt(BufferedInputStream br) throws IOException{
        if(br.read() == EXCEPTION){
            exceptions.add(getString2(br));
            return 0;
        }
        return getInt2(br);
    }
    public static long getLong(BufferedInputStream br) throws IOException{
        if(br.read() == EXCEPTION){
            exceptions.add(getString2(br));
            return 0;
        }
        return getLong2(br);
    }
    public static double getDouble(BufferedInputStream br) throws IOException{
        if(br.read() == EXCEPTION){
            exceptions.add(getString2(br));
            return 0;
        }
        return getDouble2(br);
    }

    public static String getString(BufferedInputStream br) throws IOException{
        if(br.read() == EXCEPTION){
            exceptions.add(getString2(br));
            return null;
        }
        return getString2(br);
    }
    public static int[] getIntArray(BufferedInputStream br) throws IOException{
        if(br.read() == EXCEPTION){
            exceptions.add(getString2(br));
            return null;
        }
        int N = getInt2(br);
        int[] ret = new int[N];
        for(int i = 0; i<ret.length; i++){
            ret[i] = getInt2(br);
        }
        return ret;
    }
    public static long[] getLongArray(BufferedInputStream br) throws IOException{
        if(br.read() == EXCEPTION){
            exceptions.add(getString2(br));
            return null;
        }
        int N = getInt2(br);
        long[] ret = new long[N];
        for(int i = 0; i<ret.length; i++){
            ret[i] = getLong2(br);
        }
        return ret;
    }
    public static double[] getDoubleArray(BufferedInputStream br) throws IOException{
        if(br.read() == EXCEPTION){
            exceptions.add(getString2(br));
            return null;
        }
        int N = getInt2(br);
        double[] ret = new double[N];
        for(int i = 0; i<ret.length; i++){
            ret[i] = getDouble2(br);
        }
        return ret;
    }
    public static String[] getStringArray(BufferedInputStream br) throws IOException{
        if(br.read() == EXCEPTION){
            exceptions.add(getString2(br));
            return null;
        }
        int N = getInt2(br);
        String[] ret = new String[N];
        for(int i = 0; i<ret.length; i++){
            ret[i] = getString2(br);
        }
        return ret;
    }
    
    public static boolean isNull(Object o) {
        return o == null;
    }
    public static boolean isNull(byte o) {
        return false;
    }
    public static boolean isNull(int o) {
        return false;
    }
    public static boolean isNull(long o) {
        return false;
    }
    public static boolean isNull(float o) {
        return false;
    }
    public static boolean isNull(double o) {
        return false;
    }
    public static boolean isNull(boolean o) {
        return false;
    }
    public static boolean isNull(char o) {
        return false;
    }
    public static void main(String[] args) throws Exception{
        initialize(8000);
        Socket s = getSocket();
        BufferedOutputStream bos = new BufferedOutputStream(s.getOutputStream());
        BufferedInputStream bis = new BufferedInputStream(s.getInputStream());
        System.out.println("got streams");
        startMethod(bos,1);
        bos.flush();
        System.out.println("wrote method");
        int[] wtf = new int[1000];
        wtf[0] = 0x0A0A0A0A;
        writeArg(bos, wtf);
        //bos.write(255);
        bos.flush();
        System.out.println("wrote int[]");
        System.out.println(getInt(bis));
    }
}

File: repos/app/src/csharp/LongTesterIO.cs
/*
 * Copyright (C) -2014 TopCoder Inc., All Rights Reserved.
 */

using System;
using System.IO;
using System.Text;
/**
 * I/O utility used for the communication during marathon testing of .NET solutions.
 *
 * <p>
 * Changes in version 1.1 (Module Assembly - Return Peak Memory Usage for Marathon Match - DotNet):
 * <ol>
 *      <li>Add {@link #PEAK_MEMORY_USED} constant for the peak memory data type.</li>
 *      <li>Add {@link #WritePeakMemoryUsed(BinaryWriter, long)} method to support writing
 *      of the peak memory data type.</li>
 * </ol>
 * </p>
 *
 * @author dexy
 * @version 1.1
 */
public class LongTesterIO{
    static Object resultObject;
    public const int INT_TYPE = 1;
    public const int LONG_TYPE = 2;
    public const int DOUBLE_TYPE = 3;
    public const int STRING_TYPE = 4;
    public const int VEC_INT_TYPE = 5;
    public const int VEC_LONG_TYPE = 6;
    public const int VEC_DOUBLE_TYPE = 7;
    public const int VEC_STRING_TYPE = 8;

    public const int EXCEPTION = 250;
    public const int TIMEOUT = 251;
    public const int ABORT = 252;
    public const int METHOD_START = 253;
    public const int TIME = 254;
    public const int TERMINATE = 255;
    
    /**
     * The peak memory data type.
     * @since 1.1
     */
    public const int PEAK_MEMORY_USED = 245;

    public static void WriteInt(BinaryWriter bw, int a) {
        bw.Write((byte)(a>>24));
        bw.Write((byte)(a>>16));
        bw.Write((byte)(a>>8));
        bw.Write((byte)(a&0xff));
    }
    public static void WriteLong(BinaryWriter bw, long a) {
        bw.Write((byte)(a>>56));
        bw.Write((byte)((a>>48)));
        bw.Write((byte)((a>>40)));
        bw.Write((byte)((a>>32)));
        bw.Write((byte)((a>>24)));
        bw.Write((byte)((a>>16)));
        bw.Write((byte)((a>>8)));
        bw.Write((byte)(a&0xff));
    }
    public static void WriteDouble(BinaryWriter bw, double a) {
        WriteLong(bw, BitConverter.DoubleToInt64Bits(a));
    }
    public static void WriteString(BinaryWriter bw, String a) {
        WriteInt(bw,a.Length);
System.Text.ASCIIEncoding enc = new System.Text.ASCIIEncoding();
byte[] b = enc.GetBytes(a);

        bw.Write(b);
    }

    public static void setTimeout(BinaryWriter bw, int timeout)  {
        bw.Write((byte)TIMEOUT);
        WriteInt(bw, timeout);
    }
    
    public static void startMethod(BinaryWriter bw, int method)  {
        bw.Write((byte)METHOD_START);
        WriteInt(bw, method);
    }
    
    //sends terminate command to close client while loop
    public static void terminate(BinaryWriter bw)  {
        bw.Write((byte)TERMINATE);
        Flush(bw);
        //socket.close();
    }
    
    public static void Flush(BinaryWriter bw)  {
        bw.Flush();
    }
    
    public static void WriteTime(BinaryWriter bw, int time)  {
        bw.Write((byte)TIME);
        WriteInt(bw,time);
    }

    /**
     * Writes the peak memory used (in KB) to the writer.
     *
     * @param bw binary writer where memory will be written
     * @param peakMemoryUsed the peak memory used (in KB).
     * @since 1.1
     *
     */
    public static void WritePeakMemoryUsed(BinaryWriter bw, long peakMemoryUsed) {
        bw.Write((byte) PEAK_MEMORY_USED);
        WriteLong(bw, peakMemoryUsed);
    }
    
    public static void WriteThrowable(BinaryWriter bw, Exception t)  {
        bw.Write((byte)EXCEPTION);
        WriteString(bw,t.Message+"\n"+t.StackTrace);
    }
    
    //for things like timeouts in plain text
    public static void WriteThrowable(BinaryWriter bw, String s)  {
        bw.Write((byte)EXCEPTION);
        WriteString(bw,s);
    }
    
    public static void WriteArg(BinaryWriter bw, double a) {
        bw.Write((byte)DOUBLE_TYPE);
        WriteDouble(bw,a);
    }
    public static void WriteArg(BinaryWriter bw, int a) {
        bw.Write((byte)INT_TYPE);
        WriteInt(bw,a);
    }
    public static void WriteArg(BinaryWriter bw, long a) {
        bw.Write((byte)LONG_TYPE);
        WriteLong(bw,a);
    }
    public static void WriteArg(BinaryWriter bw, String a) {
        bw.Write((byte)STRING_TYPE);
        WriteString(bw,a);
    }
    public static void WriteArg(BinaryWriter bw, int[] a) {
        bw.Write((byte)VEC_INT_TYPE);
        WriteInt(bw,a.Length);
        for(int i = 0; i<a.Length; i++){
            WriteInt(bw,a[i]);
        }
    }
    public static void WriteArg(BinaryWriter bw, long[] a) {
        bw.Write((byte)VEC_LONG_TYPE);
        WriteInt(bw,a.Length);
        for(int i = 0; i<a.Length; i++){
            WriteLong(bw,a[i]);
        }
    }
    public static void WriteArg(BinaryWriter bw, double[] a) {
        bw.Write((byte)VEC_DOUBLE_TYPE);
        WriteInt(bw,a.Length);
        for(int i = 0; i<a.Length; i++){
            WriteDouble(bw,a[i]);
        }
    }
    public static void WriteArg(BinaryWriter bw, String[] a) {
        bw.Write((byte)VEC_STRING_TYPE);
        WriteInt(bw,a.Length);
        for(int i = 0; i<a.Length; i++){
            WriteString(bw,a[i]);
        }
    }
    
    public static int GetInt2(BinaryReader br) {
        int r = br.ReadByte() << 24;
        r |= br.ReadByte()<<16;
        r |= br.ReadByte()<<8;
        r |= br.ReadByte();
        return r;
    }
    //XXX this is sucky, is string buffer faster, less mem?
static System.Text.ASCIIEncoding enc = new System.Text.ASCIIEncoding();
    public static String GetString2(BinaryReader br) {
        int Length = GetInt2(br);
byte[] b = br.ReadBytes(Length);
return enc.GetString(b);
    }
    public static long GetLong2(BinaryReader br) {
        long r = ((long)br.ReadByte()) << 56;
        r |= ((long)br.ReadByte())<<48;
        r |= ((long)br.ReadByte())<<40;
        r |= ((long)br.ReadByte())<<32;
        r |= ((long)br.ReadByte())<<24;
        r |= ((long)br.ReadByte())<<16;
        r |= ((long)br.ReadByte())<<8;
        r |= ((long)br.ReadByte());
        return r;
    }
    public static double GetDouble2(BinaryReader br) {
        return BitConverter.Int64BitsToDouble(GetLong2(br));
    }
    
    public static int GetInt(BinaryReader br) {
        if(br.ReadByte() == EXCEPTION){
            //exceptions.add(GetString2(br));
            return 0;
        }
        return GetInt2(br);
    }
    public static long GetLong(BinaryReader br) {
        if(br.ReadByte() == EXCEPTION){
            //exceptions.add(GetString2(br));
            return 0;
        }
        return GetLong2(br);
    }
    public static double GetDouble(BinaryReader br) {
        if(br.ReadByte() == EXCEPTION){
            //exceptions.add(GetString2(br));
            return 0;
        }
        return GetDouble2(br);
    }

    public static String GetString(BinaryReader br) {
        if(br.ReadByte() == EXCEPTION){
            //exceptions.add(GetString2(br));
            return null;
        }
        return GetString2(br);
    }
    public static int[] GetIntArray(BinaryReader br) {
        if(br.ReadByte() == EXCEPTION){
            //exceptions.add(GetString2(br));
            return null;
        }
        int N = GetInt2(br);
        int[] ret = new int[N];
        for(int i = 0; i<ret.Length; i++){
            ret[i] = GetInt2(br);
        }
        return ret;
    }
    public static long[] GetLongArray(BinaryReader br) {
        if(br.ReadByte() == EXCEPTION){
            //exceptions.add(GetString2(br));
            return null;
        }
        int N = GetInt2(br);
        long[] ret = new long[N];
        for(int i = 0; i<ret.Length; i++){
            ret[i] = GetLong2(br);
        }
        return ret;
    }
    public static double[] GetDoubleArray(BinaryReader br) {
        if(br.ReadByte() == EXCEPTION){
            //exceptions.add(GetString2(br));
            return null;
        }
        int N = GetInt2(br);
        double[] ret = new double[N];
        for(int i = 0; i<ret.Length; i++){
            ret[i] = GetDouble2(br);
        }
        return ret;
    }
    public static String[] GetStringArray(BinaryReader br) {
        if(br.ReadByte() == EXCEPTION){
            //exceptions.add(GetString2(br));
            return null;
        }
        int N = GetInt2(br);
        String[] ret = new String[N];
        for(int i = 0; i<ret.Length; i++){
            ret[i] = GetString2(br);
        }
        return ret;
    }
    public static bool IsNull(Object o) {
        return o == null;
    }
    public static bool IsNull(byte o) {
        return false;
    }
    public static bool IsNull(int o) {
        return false;
    }
    public static bool IsNull(long o) {
        return false;
    }
    public static bool IsNull(float o) {
        return false;
    }
    public static bool IsNull(double o) {
        return false;
    }
    public static bool IsNull(bool o) {
        return false;
    }
    public static bool IsNull(char o) {
        return false;
    }
    
}

